ocaml/xapi/import.ml

```diff
@@ -1365,80 +1369,90 @@ end
 module VBD : HandlerTools = struct
   type precheck_t = Found_VBD of API.ref_VBD | Skip | Create of API.vBD_t
 
-  let precheck __context config rpc session_id state x =
+  let precheck __context config _rpc _session_id state x =
     let vbd_record = API.vBD_t_of_rpc x.snapshot in
     let get_vbd () =
-      Client.VBD.get_by_uuid ~rpc ~session_id ~uuid:vbd_record.API.vBD_uuid
+      (* Use an internal DB call - this avoids raising an exception and logging
+         the backtrace internally in case of a (reasonably expected) absence of
+         the object with this UUID *)
+      Db.VBD.get_by_uuid_opt ~__context ~uuid:vbd_record.API.vBD_uuid
     in
-    let vbd_exists () =
-      try
-        ignore (get_vbd ()) ;
-        true
-      with _ -> false
+    let vbd_opt =
+      (* If there's already a VBD with the same UUID and we're preserving UUIDs, use that one. *)
+      if config.full_restore then
+        get_vbd ()
+      else
+        None
     in
-    if config.full_restore && vbd_exists () then
-      let vbd = get_vbd () in
-      Found_VBD vbd
-    else
-      let vm =
-        log_reraise
-          ("Failed to find VBD's VM: " ^ Ref.string_of vbd_record.API.vBD_VM)
-          (lookup vbd_record.API.vBD_VM)
-          state.table
-      in
-      (* If the VBD is supposed to be attached to a PV guest (which doesn't support
-         				 currently_attached empty drives) then throw a fatal error. *)
-      let original_vm =
-        get_vm_record
-          (find_in_export (Ref.string_of vbd_record.API.vBD_VM) state.export)
-      in
-      (* Note: the following is potentially inaccurate: the find out whether a running or
-       * suspended VM has booted HVM, we must consult the VM metrics, but those aren't
-       * available in the exported metadata. *)
-      let has_qemu = Helpers.will_have_qemu_from_record original_vm in
-      (* In the case of dry_run live migration, don't check for
-         				 missing disks as CDs will be ejected before the real migration. *)
-      let dry_run, live =
-        match config.import_type with
-        | Metadata_import {dry_run; live; _} ->
-            (dry_run, live)
-        | _ ->
-            (false, false)
-      in
-      ( if
-          vbd_record.API.vBD_currently_attached
-          && not (exists vbd_record.API.vBD_VDI state.table)
-        then
-          (* It's only ok if it's a CDROM attached to an HVM guest, or it's part of SXM and we know the sender would eject it. *)
-          let will_eject =
-            dry_run && live && original_vm.API.vM_power_state <> `Suspended
-          in
-          if not (vbd_record.API.vBD_type = `CD && (has_qemu || will_eject))
+    match vbd_opt with
+    | Some vbd ->
+        Found_VBD vbd
+    | None -> (
+        warn
+          "Did not find an already existing VBD with the same uuid=%s, try to \
```

Comment:
Or we didn't look for an existing one, if this isn't a full restore. Also I'd call this `info`, not `warn`.

Reply:
"Full restore" is an odd name when it really means "preserve uuids"...